pageserver/src/page_cache.rs

```diff
@@ -0,0 +1,591 @@
+    //
+    // Currently, the page cache only stores materialized page images. In the
+    // future, to add support for a new kind of "thing" to cache, you will need to
+    // add public interface routines above, and code to deal with the "mappings"
+    // after this section. But the routines in this section should not require changes.
+
+    /// Look up a page in the cache.
+    ///
+    /// If the search criteria is not exact, *cache_key is updated with the key for exact
+    /// key of the returned page. (For materialized pages, that means that the LSN
+    /// in 'cache_key' is updated with the LSN of the returned page version.)
+    ///
+    /// If no page is found, returns None and *cache_key is left unmodified.
+    ///
+    fn try_lock_for_read(&self, cache_key: &mut CacheKey) -> Option<PageReadGuard> {
+        let cache_key_orig = cache_key.clone();
+        if let Some(slot_idx) = self.search_mapping(cache_key) {
+            // The page was found in the mapping. Lock the slot, and re-check
+            // that it's still what we expected (because we don't released the mapping
+            // lock already, another thread could have evicted the page)
+            let slot = &self.slots[slot_idx];
+            let inner = slot.inner.read().unwrap();
+            if inner.key.as_ref() == Some(cache_key) {
+                slot.inc_usage_count();
+                return Some(PageReadGuard(inner));
+            } else {
+                // search_mapping might have modified the search key; restore it.
+                *cache_key = cache_key_orig;
+            }
+        }
+        None
+    }
+
+    /// Return a locked buffer for given block.
+    ///
+    /// Like try_lock_for_read(), if the search criteria is not exact and the page is
+    /// already found in the cache, *cache_key is updated.
+    ///
+    /// If the page is not found in the cache, this allocates a new buffer for it. The
+    /// caller may then initialize the buffer with the contents, and call mark_valid().
+    ///
+    /// Example usage:
+    ///
+    /// match cache.lock_for_read(&key) {
+    ///     ReadBufResult::Found(read_guard) => {
+    ///         // The page was found in cache. Use it
+    ///         ...
+    ///     },
+    ///     ReadBufResult::NotFound(write_guard) => {
+    ///         // The page was not found in cache. Read it from disk into the buffer
+    ///         read_from_disk(write_guard);
+    ///         // The buffer contents are now valid. Tell the page cache.
+    ///         write_guard.mark_valid();
+    ///     },
+    /// }
+    ///
+    #[allow(unused)] // this is currently unused
+    fn lock_for_read(&self, cache_key: &mut CacheKey) -> ReadBufResult {
+        loop {
+            // First check if the key already exists in the cache.
+            if let Some(read_guard) = self.try_lock_for_read(cache_key) {
+                return ReadBufResult::Found(read_guard);
+            }
```

Comment:
Can we move this outside the loop to fix the todo in the `try_insert_mapping` guard below? Same for lock_for_write

Reply:
That's not quite enough, then we would get into an infinite loop, finding a new victim, and finding that the mapping is still there on every iteration.

So we still need to do the inner part of try_lock_for_read() here, but we could skip the search_mapping() call after the first iteration, and check the 'existing_slot_idx' that try_insert_mapping() returned.